File: Treachery-iOS/Treachery-iOS/Models/PlanechaseState.swift
//
//  PlanechaseState.swift
//  Treachery-iOS
//
//  Created by Luke Solomon on 3/16/26.
//

import Foundation

struct PlanechaseState: Codable, Hashable {
    var useOwnDeck: Bool
    var currentPlaneId: String?
    var usedPlaneIds: [String]
    var lastDieRollerId: String?
    var dieRollCount: Int
    var chaoticAetherActive: Bool
    var secondaryPlaneId: String?

    enum CodingKeys: String, CodingKey {
        case useOwnDeck = "use_own_deck"
        case currentPlaneId = "current_plane_id"
        case usedPlaneIds = "used_plane_ids"
        case lastDieRollerId = "last_die_roller_id"
        case dieRollCount = "die_roll_count"
        case chaoticAetherActive = "chaotic_aether_active"
        case secondaryPlaneId = "secondary_plane_id"
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        useOwnDeck = try container.decodeIfPresent(Bool.self, forKey: .useOwnDeck) ?? false
        currentPlaneId = try container.decodeIfPresent(String.self, forKey: .currentPlaneId)
        usedPlaneIds = try container.decodeIfPresent([String].self, forKey: .usedPlaneIds) ?? []
        lastDieRollerId = try container.decodeIfPresent(String.self, forKey: .lastDieRollerId)
        dieRollCount = try container.decodeIfPresent(Int.self, forKey: .dieRollCount) ?? 0
        chaoticAetherActive = try container.decodeIfPresent(Bool.self, forKey: .chaoticAetherActive) ?? false
        secondaryPlaneId = try container.decodeIfPresent(String.self, forKey: .secondaryPlaneId)
    }

    init(
        useOwnDeck: Bool,
        currentPlaneId: String?,
        usedPlaneIds: [String],
        lastDieRollerId: String?,
        dieRollCount: Int,
        chaoticAetherActive: Bool = false,
        secondaryPlaneId: String? = nil
    ) {
        self.useOwnDeck = useOwnDeck
        self.currentPlaneId = currentPlaneId
        self.usedPlaneIds = usedPlaneIds
        self.lastDieRollerId = lastDieRollerId
        self.dieRollCount = dieRollCount
        self.chaoticAetherActive = chaoticAetherActive
        self.secondaryPlaneId = secondaryPlaneId
    }
}

File: TreacheryAndroid/app/src/main/java/com/solomon/treachery/model/PlanechaseState.kt
package com.solomon.treachery.model

data class PlanechaseState(
    val useOwnDeck: Boolean = false,
    val currentPlaneId: String? = null,
    val usedPlaneIds: List<String> = emptyList(),
    val lastDieRollerId: String? = null,
    val dieRollCount: Int = 0,
    val chaoticAetherActive: Boolean = false,
    val secondaryPlaneId: String? = null
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "use_own_deck" to useOwnDeck,
        "current_plane_id" to currentPlaneId,
        "used_plane_ids" to usedPlaneIds,
        "last_die_roller_id" to lastDieRollerId,
        "die_roll_count" to dieRollCount,
        "chaotic_aether_active" to chaoticAetherActive,
        "secondary_plane_id" to secondaryPlaneId
    )

    companion object {
        fun fromMap(data: Map<String, Any?>): PlanechaseState = PlanechaseState(
            useOwnDeck = data["use_own_deck"] as? Boolean ?: false,
            currentPlaneId = data["current_plane_id"] as? String,
            usedPlaneIds = (data["used_plane_ids"] as? List<*>)?.filterIsInstance<String>() ?: emptyList(),
            lastDieRollerId = data["last_die_roller_id"] as? String,
            dieRollCount = (data["die_roll_count"] as? Number)?.toInt() ?: 0,
            chaoticAetherActive = data["chaotic_aether_active"] as? Boolean ?: false,
            secondaryPlaneId = data["secondary_plane_id"] as? String
        )
    }
}
